doma-core/src/main/java/org/seasar/doma/jdbc/criteria/query/BuilderSupport.java

```diff
@@ -62,6 +63,40 @@ public BuilderSupport(
         new UserDefinedExpressionDeclarationItemVisitor();
   }
 
+  public void with(List<WithContext> withContexts) {
+    if (withContexts.isEmpty()) return;
+
+    buf.appendSql("with ");
+    for (WithContext withContext : withContexts) {
+      var entityMetamodel = withContext.entityMetamodel();
+      var setOperationContext = withContext.setOperationContext();
+      // cte name
+      buf.appendSql(
+          entityMetamodel
+              .asType()
+              .getQualifiedTableName(config.getNaming()::apply, config.getDialect()::applyQuote));
+      // cte column list
+      buf.appendSql("(");
+      for (PropertyMetamodel<?> propertyMetamodel : entityMetamodel.allPropertyMetamodels()) {
+        String columnName =
+            propertyMetamodel
+                .asType()
+                .getColumnName(config.getNaming()::apply, config.getDialect()::applyQuote);
+        buf.appendSql(columnName);
+        buf.appendSql(", ");
+      }
+      buf.cutBackSql(2);
+      buf.appendSql(") as (");
+      // cte body
+      SetOperationBuilder builder =
+          new SetOperationBuilder(config, setOperationContext, commenter, buf, aliasManager);
+      builder.build();
+      buf.appendSql("), ");
+    }
+    buf.cutBackSql(2);
+    buf.appendSql(" ");
+  }
```

Comment:
possible to common with the following code.
but desirable explicitly CTE column.
https://github.com/domaframework/doma/blob/add3be4dcdc8f4363b23fcbb79c6a8d5aa7093bc/doma-core/src/main/java/org/seasar/doma/jdbc/criteria/query/BuilderSupport.java#L100-L136

I think better A than B

A
```sql
with dcCte1(DEPARTMENT_ID, COUNT) as (select t3_.DEPARTMENT_ID, count(t4_.ADDRESS_ID)
                                      from DEPARTMENT t3_
                                               left outer join EMPLOYEE t4_ on (t3_.DEPARTMENT_ID = t4_.DEPARTMENT_ID)
                                      group by t3_.DEPARTMENT_ID),
```

B
```sql
with dcCte1 as (select t3_.DEPARTMENT_ID as DEPARTMENT_ID, count(t4_.ADDRESS_ID) as COUNT
                                      from DEPARTMENT t3_
                                               left outer join EMPLOYEE t4_ on (t3_.DEPARTMENT_ID = t4_.DEPARTMENT_ID)
                                      group by t3_.DEPARTMENT_ID),
```